### PR TITLE
Applies the changes done in host functions API.

### PR DESCRIPTION
**Pull request by rimbi**
## Purpose

Applies the changes in the host functions/chain API.

## Checklist

- [x ] My code follows the style of this project.
- [ x] The code compiles without warnings.
- [ x] I have performed a self-review of the changes.
- [ x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x ] I accept the above linked CLA.

### CODE REVIEW

File: concordium-std/src/impls.rs [hunk 1]
```diff
@@ -575,61 +576,76 @@ fn prepend_generic_map_key<K: Serial>(key: K) -> Vec<u8> {
...
-        let prefix_start = prefix.as_ptr();
-        let prefix_len = prefix.len() as u32; // Wasm usize == 32bit.
-        let iterator_id = unsafe { iterator(prefix_start, prefix_len) };
-        ContractStateIter {
-            iterator_id,
+    fn iterator(
+        &self,
+        prefix: &[u8],
+    ) -> Result<<T as ContractStateApi>::IterType, ContractStateError> {
+        let locked = self.locked_subtrees.borrow().keys().any(|p| {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:633]: The contract should not be checking whether a subtree is locked on its own. 
It should just call `iterator()` on the host and parse the response.

Comment (Bargsteen) [reply concordium-std/src/impls.rs:633]: But the test trie  _should_ perform these kinds of checks so that the semantics are identical to those of the host.

File: concordium-std/src/impls.rs [hunk 2]
```diff
@@ -697,21 +716,18 @@ where
     }
 
     pub fn is_empty(&self) -> bool {
-        if let Some(_) = self.state_ll.borrow().iterator(&self.prefix).next() {
-            false
-        } else {
-            true
-        }
+        let iterator = self.state_ll.borrow().iterator(&self.prefix);
+        iterator.is_err() || iterator.unwrap().next().is_none()
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:720]: `iterator(..)` will return an error if you have more than `N` iterators for the same key. 
So if you get `N` iterators for `self.prefix` and then call `is_empty()`, it will return `true` because you will get an error (despite the trie not being empty).

File: concordium-std/src/test_infrastructure/trie.rs [hunk 3]
```diff
@@ -102,15 +74,179 @@ impl StateTrie {
     }
 }
 
-#[derive(Debug)]
+impl ContractStateApi for StateTrie {
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure/trie.rs:77]: I don't think this makes sense to implement for StateTrie, as it cannot provide a complete host API.

File: concordium-std/src/impls.rs [hunk 4]
```diff
@@ -1606,3 +1623,200 @@ where
         source.read_u64().and_then(|set_prefix| Ok(StateSet::open(Rc::clone(state), set_prefix)))
     }
 }
+
+impl ContractStateApi for () {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:1627]: Given my comment on the `impl ContractStateApi for StateTrie`, I am not sure this makes so much sense to have.
The idea of wrapping the unsafe calls is fine, but if we do that, we should rewrite the other implementations (those not related to state) to also use this API. For example `slot_time()` in `ChainMetaExtern`. But changing those calls should not be part of the new state PR. (It's large enough already ;) )

File: concordium-std/src/errors.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+//! This internal module provides the error types used in the crate.
+
+#[derive(Debug)]
+pub enum ContractStateError {
```
Comment (Bargsteen) [inline concordium-std/src/errors.rs:4]: We already have a number of error types in `types.rs`, so I think this should live there as well 😊 
If you think that the errors should live in a separate error module, then we can move all of them in a separate PR at some point.

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Good work in general. I like that you added tests for the iterators 👍 

There are a few issues, the primary one being that `ContractStateLL` should not keep track of the locked subtrees. Instead, it should rely on the host doing this job.

I'm likely to blame for this. I should have made that clear when we talked about the task. Luckily, you should already have most of the parts needed. They should just be changed so the subtree-lock checks are only performed on the test trie. 😊

File: concordium-std/src/errors.rs [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    /// The subtree is locked.
+    SubtreeLocked,
+    /// The iterator does not exist.
```
Comment (abizjak) [inline concordium-std/src/errors.rs:7]: The name of the variant does not match the comment above it.

File: concordium-std/src/errors.rs (hunk 5 above)
Comment (abizjak) [reply concordium-std/src/errors.rs:4]: I agree with Kasper. Do not partially move errors here. This belongs in types.rs.

File: concordium-std/src/impls.rs [hunk 7]
```diff
@@ -145,10 +146,10 @@ impl HasContractStateEntry for StateEntry {
     }
 
     fn get_key(&self) -> Vec<u8> {
```
Comment (abizjak) [inline concordium-std/src/impls.rs:148]: I don't think this method should return a `Vec<u8>`. The principle we have until now is that it is in principle possible to go without an allocator. So our API should as much as practical not rely on an allocator being present.

Instead the buffer should be provided by the caller.

File: concordium-std/src/impls.rs [hunk 8]
```diff
@@ -575,61 +576,76 @@ fn prepend_generic_map_key<K: Serial>(key: K) -> Vec<u8> {
...
-    type ContractStateData = ();
-    type EntryType = StateEntry;
-    type IterType = ContractStateIter;
+impl<T> ContractStateLL<T>
+where
+    T: ContractStateApi,
+{
+    pub fn new(state_api: T) -> Self {
+        Self {
+            locked_subtrees: Default::default(),
```
Comment (abizjak) [inline concordium-std/src/impls.rs:585]: I don't think the std should keep track of locked trees. THe host already does that.

File: concordium-std/src/traits.rs [hunk 9]
```diff
@@ -321,6 +340,163 @@ pub trait DeserialCtx: Sized {
...
 where
-    S: HasContractStateLL, {
+    S: HasContractStateLL,
+{
     fn deserial_state_ctx<R: Read>(state: &Rc<RefCell<S>>, source: &mut R) -> ParseResult<Self>;
 }
+
+/// A trait for the contract state / host functions API.
+/// It provides abstractions to access the state of the contract, and to invoke host functions.
+pub trait ContractStateApi: Debug {
```
Comment (abizjak) [inline concordium-std/src/traits.rs:350]: What is the point of this trait? It is just adding a trait that names the functions that are imported. What problem does this solve?

File: concordium-std/src/types.rs [hunk 10]
```diff
@@ -368,7 +375,7 @@ macro_rules! claim_ne {
 ///
 /// # Example
 /// Defining a custom error type
-/// ```rust
+/// ```ignore
```
Comment (abizjak) [inline concordium-std/src/types.rs:378]: This should generally not be done. It makes docs less nice to look at (e.g., worse syntax highlighting). 

THe correct way to fix this is to make it compile, but that should not be part of this PR.

Please revert this.

File: examples/new-state/src/lib.rs [hunk 11]
```diff
@@ -167,7 +167,7 @@ mod tests {
     fn test_init() {
         let ctx = InitContextTest::empty();
 
-        let state_ll = Rc::new(RefCell::new(ContractStateLLTest::new()));
+        let state_ll = Rc::new(RefCell::new(ContractStateLLTest::default()));
```
Comment (abizjak) [inline examples/new-state/src/lib.rs:170]: I strongly prefer being explicit instead of implicit, and default is implicit. Is there are particular reason for this change?

File: examples/new-state/src/lib.rs [hunk 12]
```diff
@@ -215,7 +215,7 @@ mod tests {
             Some(Ok(expected_token_count))
         );
 
-        let mut a_set_iter = state_reloaded.another_struct.a_set.iter();
+        let mut a_set_iter = state_reloaded.another_struct.a_set.iter().unwrap();
```
Comment (abizjak) [inline examples/new-state/src/lib.rs:218]: Use `expect` with a meaningful explanation so that if it fails a useful message is displayed.

File: concordium-std/src/traits.rs (hunk 9 above)
Comment (abizjak) [reply concordium-std/src/traits.rs:350]: Since we did not need it until now, I do not think it is needed now for contract state changes either.

File: concordium-std/src/traits.rs [hunk 13]
```diff
@@ -133,33 +135,50 @@ where
     fn get_key(&self) -> Vec<u8>;
 }
 
-pub trait HasContractStateLL {
-    type ContractStateData;
-    type EntryType: HasContractStateEntry;
-    type IterType: Iterator<Item = Self::EntryType>;
+pub trait HasContractStateLL: Debug {
```
Comment (abizjak) [inline concordium-std/src/traits.rs:138]: Why the debug supertrait? 

Code size is an issue, so we should not require things in smart contracts that are not crucial.

File: concordium-std/src/traits.rs [hunk 14]
```diff
@@ -2,12 +2,14 @@
 //! This allows setting-up mock objects for testing individual
 //! contract invocations.
 
-use std::{cell::RefCell, rc::Rc};
+use std::{cell::RefCell, fmt::Debug, rc::Rc};
 
 #[cfg(not(feature = "std"))]
 use alloc::vec::Vec;
```
Comment (abizjak) [inline concordium-std/src/traits.rs:9]: No spaces between imports please. And let `fmt` sort them.

File: concordium-std/src/traits.rs [hunk 15]
```diff
@@ -2,12 +2,14 @@
 //! This allows setting-up mock objects for testing individual
 //! contract invocations.
 
-use std::{cell::RefCell, rc::Rc};
+use std::{cell::RefCell, fmt::Debug, rc::Rc};
```
Comment (abizjak) [inline concordium-std/src/traits.rs:6]: And here.

File: concordium-std/src/test_infrastructure/trie.rs [hunk 16]
```diff
@@ -254,7 +387,7 @@ impl Node {
 mod tests {
     use concordium_contracts_common::{to_bytes, Deserial, Write};
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure/trie.rs:389]: No spaces between imports please. And let `fmt` sort them out.

File: concordium-std/src/test_infrastructure/trie.rs (hunk 3 above)
Comment (abizjak) [reply concordium-std/src/test_infrastructure/trie.rs:77]: I agree. `ContractStateApi` trait should be removed.

File: concordium-std/src/test_infrastructure.rs [hunk 17]
```diff
@@ -971,7 +934,8 @@ mod test {
             false
         );
 
-        let mut iter = state.get::<_, StateSet<u8, _>>(my_set_key).unwrap().unwrap().iter();
+        let mut iter =
+            state.get::<_, StateSet<u8, _>>(my_set_key).unwrap().unwrap().iter().unwrap();
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure.rs:938]: Please use `expect` and explain why unwrapping is intended. Otherwise the error message will be hard to understand.

File: concordium-std/src/test_infrastructure.rs [hunk 18]
```diff
@@ -884,7 +847,7 @@ mod test {
     fn high_level_insert_get() {
         let expected_value: u64 = 123123123;
         let mut state: ContractStateHL<ContractStateLLTest> =
-            HasContractStateHL::open(Rc::new(RefCell::new(ContractStateLLTest::new())));
+            HasContractStateHL::open(Rc::new(RefCell::new(ContractStateLLTest::default())));
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure.rs:850]: I prefer being explicit, so I don't see a need to use default instead of new.

File: concordium-std/src/test_infrastructure.rs [hunk 19]
```diff
@@ -772,6 +725,16 @@ impl Seek for StateEntryTest {
     }
 }
 
+mod trie;
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure.rs:728]: I think this is an unusal way to organize modules and is not the way we use generally.

When using nested submodules they are either in the same file, or we use a `mod.rs` file that names the root.

File: concordium-std/src/test_infrastructure.rs [hunk 20]
```diff
@@ -8,7 +8,7 @@
 //!
 //! # Example
 //!
-//! ```rust
+//! ```ignore
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure.rs:11]: Please do not ignore. This will make `cargo doc` worse to read.

The correct way to fix this is to make it compile, but there is a separate issue for that #77 and should be done in a separate PR.

So please revert all of these.

File: concordium-std/src/impls.rs (hunk 4 above)
Comment (abizjak) [reply concordium-std/src/impls.rs:1627]: I don't think the idea makes sense. I do not see what value this abstraction would provide. Why is it valuable to provide a function with the same name that just wraps the unsafe? 

The functions are unsafe for a reason, and rely on correct usage, which is not ensured by this implementation. So these functions are in fact very unsafe and can lead to arbitrary memory corruption. For example `get_receive_sender` must have a buffer that is of correct size, which is not ensured here.

**Review by abizjak — CHANGES_REQUESTED** [review]

File: concordium-std/src/impls.rs [hunk 21]
```diff
@@ -638,21 +691,18 @@ where
...
-        }
+        let iterator = self.state_ll.borrow().iterator(&self.prefix);
+        iterator.is_err() || iterator.unwrap().next().is_none()
     }
 
-    pub fn iter(&self) -> StateMapIter<K, V, S> {
-        let state_iter = self.state_ll.borrow().iterator(&self.prefix);
-        StateMapIter {
+    pub fn iter(&self) -> Result<StateMapIter<K, V, S>, ContractStateError> {
+        let state_iter = self.state_ll.borrow().iterator(&self.prefix)?;
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:703]: `StateMapIter` should keep a reference to the `StateMap` so that this method can't be called twice without dropping the `StateMapIter`. When that is implemented, we can unwrap this call to `iterator`.

File: concordium-std/src/impls.rs [hunk 22]
```diff
@@ -680,19 +727,23 @@ where
     fn next(&mut self) -> Option<Self::Item> {
         self.state_iter.next().and_then(|mut entry| {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:728]: This is where we should get the key from the `state_iter` (mentioned previously).

I am not sure what the agreement is regarding `state_iterator_key_read`.
From `prims.rs` it seems like it should be called _before_ `next()` to get the correct key.
But if that is the case, then it `state_iterator_key_read` should also be able to return an error, if there is no next. (Or we should peek first). What do you think @abizjak ?

File: concordium-std/src/impls.rs [hunk 23]
```diff
@@ -911,29 +963,26 @@ where
...
     fn read(&mut self, buf: &mut [u8]) -> ParseResult<usize> {
-        let len: u32 = {
-            match buf.len().try_into() {
-                Ok(v) => v,
-                _ => return Err(ParseError::default()),
-            }
-        };
-        let num_read =
-            unsafe { get_parameter_section(0, buf.as_mut_ptr(), len, self.current_position) };
+        let num_read = self.get_parameter_section(0, buf, self.current_position).unwrap_abort();
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:985]: Please don't make unrelated changes such as this. It makes it harder and more time-consuming to review. Aim for minimal changes that _only_ relate to the specific PR 😊

File: concordium-std/src/impls.rs [hunk 24]
```diff
@@ -942,36 +991,30 @@ impl Read for ExternParameter {
...
-        let num_read = unsafe {
-            get_parameter_section(self.i.into(), buf.as_mut_ptr(), len, self.current_position)
-        };
-        if num_read >= 0 {
-            self.current_position += num_read as u32;
-            Ok(num_read as usize)
-        } else {
-            Err(ParseError::default())
-        }
+        let num_read = self.get_parameter_section(self.i.into(), buf, self.current_position).map_err(|_| ParseError::default())?;
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:1005]: Same here

File: concordium-std/src/impls.rs [hunk 25]
```diff
@@ -942,36 +991,30 @@ impl Read for ExternParameter {
 impl HasParameter for ExternParameter {
     #[inline(always)]
     // parameter 0 always exists so this is correct
-    fn size(&self) -> u32 { unsafe { get_parameter_size(0) as u32 } }
+    fn size(&self) -> u32 {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:994]: And here

File: concordium-std/src/impls.rs [hunk 26]
```diff
@@ -942,36 +991,30 @@ impl Read for ExternParameter {
...
 }
 
+impl HasFunctionsAPIWrapper for CallResponse {}
+
 /// CallResponse can only be constured in this crate. As a result whenever it is
 /// constructed it will point to a valid parameter, which means that
 /// `get_parameter_size` will always return a non-negative value.
 impl HasCallResponse for CallResponse {
-    fn size(&self) -> u32 { unsafe { get_parameter_size(self.i.get()) as u32 } }
+    fn size(&self) -> u32 { self.get_parameter_size(self.i.get()).unwrap_abort() }
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:1017]: And here

File: concordium-std/src/impls.rs [hunk 27]
```diff
@@ -778,7 +830,7 @@ where
     S: HasContractStateLL,
 {
     fn store(self, prefix: &[u8], state_ll: Rc<RefCell<S>>) {
-        match state_ll.borrow_mut().entry(prefix) {
+        match state_ll.borrow_mut().entry(prefix).unwrap_abort() {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:833]: We need to make sure this won't cause problems.

File: concordium-std/src/impls.rs [hunk 28]
```diff
@@ -762,13 +813,14 @@ where
         key_with_prefix
     }
 
-    pub fn iter(&self) -> StateSetIter<T, S> {
-        let state_iter = self.state_ll.borrow().iterator(&self.prefix);
-        StateSetIter {
+    pub fn iter(&self) -> Result<StateSetIter<T, S>, ContractStateError> {
```
Comment (Bargsteen) [inline concordium-std/src/impls.rs:816]: Same idea as with the iter for statemap. It should have a reference to the stateset.

File: concordium-std/src/test_infrastructure/trie.rs [hunk 29]
```diff
@@ -100,20 +81,77 @@ impl StateTrie {
...
+    }
+
+    pub fn delete_iterator(&mut self, iterator: Iter) -> bool {
+        if iterator.clone_count() > 2 {
+            println!("clone count = {}", iterator.clone_count());
+            return true;
+        }
+        println!("2");
+        let initial_count = self.iterators.borrow().len();
+        self.iterators.borrow_mut().retain(|_, iter| iter != &iterator);
```
Comment (Bargsteen) [inline concordium-std/src/test_infrastructure/trie.rs:144]: An optimization here would be that the iterator keeps track of its own prefix, so we can use `remove(iterator.key)` instead. But the current solution is probably fine for testing.

**Review by Bargsteen — CHANGES_REQUESTED** [review]

File: concordium-std/src/impls.rs [hunk 30]
```diff
@@ -578,12 +633,7 @@ impl Iterator for ContractStateIter {
     type Item = StateEntry;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let res = unsafe { next(self.iterator_id) };
-        if res < 0 {
-            None
-        } else {
-            Some(StateEntry::open(res as u32))
-        }
+        self.state_iterator_next().unwrap_abort()
```
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-std/src/impls.rs:634]: That depends a bit on what the intent of this `Iterator` implementation is.

One option is to abort, but this seems like the worst option.
Another option is to just return None, terminate the iterator when encountering an error.

The last option is to have an iterator over `Result<StateEntry, _>`. This would then reflect the fact that this is just a convenience wrapper around `state_iterator_next`.

I would imagine this is only used in low-level implementations, so my preference would be option (3), or remove the implementation of the Iterator trait entirely.

File: concordium-std/src/impls.rs [hunk 31]
```diff
@@ -606,17 +656,17 @@ where
 
     pub fn get(&self, key: K) -> Option<V> {
         let k = self.key_with_map_prefix(&key);
-        self.state_ll.borrow().lookup(&k).and_then(|mut entry| {
-            Some(V::deserial_state_ctx(&self.state_ll, &mut entry).expect_report(
+        self.state_ll.borrow().lookup(&k).map(|mut entry| {
+            V::deserial_state_ctx(&self.state_ll, &mut entry).expect_report(
```
Comment (abizjak) [inline concordium-std/src/impls.rs:656]: `expect_report` **cannot be used in production code**.

It is only available when testing locally and you **will not be able to build a contract that uses it**. So its use must be removed.

File: concordium-std/src/impls.rs (hunk 2 above)
Comment (abizjak) [reply concordium-std/src/impls.rs:695]: Since `N=2^16` perhaps we should just abort in that case?

Comment (abizjak) [reply concordium-std/src/impls.rs:695]: Looking at the details of the implementation in the host now, there seems to be no disadvantage in using a u32 for this limit, so I think we can raise it to 2^32, which makes using abort here appropriate.

File: concordium-std/src/impls.rs (hunk 22 above)
Comment (abizjak) [reply concordium-std/src/impls.rs:724]: Yes, you are right. `state_iterator_key_read` should return an Option indicating whether there is a next value or not.
I'll revise the host implementation and specification.

This also makes it cheaper to use to check whether an iterator is empty (i.e., `is_empty` method above for the state map).

File: concordium-std/src/impls.rs [hunk 32]
```diff
@@ -680,19 +727,23 @@ where
...
-            Some((k, v))
+            if let Ok(key) = key {
+                let mut key_cursor = Cursor {
+                    data:   key,
+                    offset: 8, // Items in a map always start with the set prefix which is 8 bytes.
+                };
+                let k = K::deserial(&mut key_cursor).expect_report(
+                    "Deserializing key failed. State has been incorrectly altered using the low-level \
+                     API.",
+                );
```
Comment (abizjak) [inline concordium-std/src/impls.rs:734]: As above, expect_report is not to be used in production code, only testing.

File: concordium-std/src/impls.rs [hunk 33]
```diff
@@ -680,19 +727,23 @@ where
...
+                let v = V::deserial_state_ctx(&self.state_ll, &mut entry).expect_report(
+                    "Deserializing value failed. State has been incorrectly altered using the \
+                     low-level API.",
+                );
```
Comment (abizjak) [inline concordium-std/src/impls.rs:738]: And here.

File: concordium-std/src/impls.rs (hunk 28 above)
Comment (abizjak) [reply concordium-std/src/impls.rs:812]: Yes.

I would just like you don't get bogged down in this part and that we get something working if the references and lifetimes are not the most straightforward concept.

We can do it in stages.

File: concordium-std/src/impls.rs (hunk 23 above)
Comment (abizjak) [reply concordium-std/src/impls.rs:981]: I completely agree with this and I am strongly against such changes.

There is already so much subtle stuff going on that we do not need more changes.

File: concordium-std/src/impls.rs [hunk 34]
```diff
@@ -942,36 +991,30 @@ impl Read for ExternParameter {
 impl HasParameter for ExternParameter {
     #[inline(always)]
     // parameter 0 always exists so this is correct
-    fn size(&self) -> u32 { unsafe { get_parameter_size(0) as u32 } }
+    fn size(&self) -> u32 {
+        self.get_parameter_size(0).unwrap_abort()
+    }
 }
 
+impl HasFunctionsAPIWrapper for ExternParameter {}
```
Comment (abizjak) [inline concordium-std/src/impls.rs:995]: Remove this trait. It does not serve a purpose.

File: concordium-std/src/impls.rs [hunk 35]
```diff
@@ -942,36 +991,30 @@ impl Read for ExternParameter {
...
-        } else {
-            Err(ParseError::default())
-        }
+        let num_read = self.get_parameter_section(self.i.into(), buf, self.current_position).map_err(|_| ParseError::default())?;
+        self.current_position += num_read as u32;
+        Ok(num_read as usize)
     }
 }
 
+impl HasFunctionsAPIWrapper for CallResponse {}
```
Comment (abizjak) [inline concordium-std/src/impls.rs:1007]: I don't get these implementations. `CallResponse` should only have Read + Seek implementation. That is the only thing you can do with it.

File: concordium-std/src/impls.rs [hunk 36]
```diff
@@ -1563,7 +1606,12 @@ impl<A> UnwrapAbort for Option<A> {
     type Unwrap = A;
 
     #[inline(always)]
-    fn unwrap_abort(self) -> Self::Unwrap { self.unwrap_or_else(|| crate::trap()) }
+    fn unwrap_abort(self) -> Self::Unwrap {
+        match self {
+            Some(x) => x,
+            None => crate::trap(),
+        }
+    }
```
Comment (abizjak) [inline concordium-std/src/impls.rs:1610]: Why was this change made?

Please avoid unrelated changes unless they are bugs. In which case it is still better to document a bug and fix it in a separate PR.

File: concordium-std/src/test_infrastructure/trie.rs [hunk 37]
```diff
@@ -17,6 +17,16 @@ pub struct StateTrie {
     nodes:         Node,
     next_entry_id: Cell<StateEntryId>,
     entry_map:     RefCell<HashMap<StateEntryId, Vec<Index>>>,
+    iterators:     RefCell<BTreeMap<Vec<usize>, Iter>>,
+}
+
+/// Default constructor for a new state trie.
+/// 
+/// Added after clippy warning.
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure/trie.rs:25]: That's not a good comment :) It does not provide value to the reader.

Instead document what the default value is, i.e., the empty state trie.

File: concordium-std/src/test_infrastructure/trie.rs [hunk 38]
```diff
@@ -100,20 +81,77 @@ impl StateTrie {
...
+        self.iterators.borrow_mut().insert(index_prefix, iter.clone());
+        Ok(iter)
+    }
+
+    pub fn delete_iterator(&mut self, iterator: Iter) -> bool {
+        if iterator.clone_count() > 2 {
+            println!("clone count = {}", iterator.clone_count());
+            return true;
+        }
+        println!("2");
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure/trie.rs]: I guess debug print? Should probably be removed.

File: concordium-std/src/test_infrastructure/trie.rs [hunk 39]
```diff
@@ -100,20 +81,77 @@ impl StateTrie {
...
+        }
+        let node = self.nodes.lookup_node(&index_prefix).ok_or(ContractStateError::EntryNotFound)?;
+        let iter = Iter::new(self, index_prefix.clone(), node);
+        self.iterators.borrow_mut().insert(index_prefix, iter.clone());
+        Ok(iter)
+    }
+
+    pub fn delete_iterator(&mut self, iterator: Iter) -> bool {
+        if iterator.clone_count() > 2 {
+            println!("clone count = {}", iterator.clone_count());
```
Comment (abizjak) [inline concordium-std/src/test_infrastructure/trie.rs]: I guess debug print? Should probably be removed.

File: concordium-std/src/types.rs [hunk 40]
```diff
@@ -514,3 +514,21 @@ pub(crate) mod sealed {
...
+    /// The entry does not exist.
+    EntryNotFound,
+    /// The parameter does not exist.
+    ParameterNotFound,
+    /// The specified size is too big.
+    SizeTooLarge,
+    /// Overflow error.
+    Overflow,
+    /// The iterator limit for a prefix has been reached.
+    IteratorLimitForPrefixExceeded,
```
Comment (abizjak) [inline concordium-std/src/types.rs:533]: I suppose you also want to add `IteratorNotFound` error here.

File: concordium-std/src/types.rs [hunk 41]
```diff
@@ -514,3 +514,21 @@ pub(crate) mod sealed {
...
+    /// The operation is not allowed.
+    NotAllowed,
```
Comment (abizjak) [inline concordium-std/src/types.rs:535]: Where is this used? At a quick glance I could not see it.

File: examples/new-state/src/lib.rs [hunk 42]
```diff
@@ -208,7 +208,7 @@ mod tests {
             Some(expected_token_count)
         );
 
-        let mut a_set_iter = state_reloaded.another_struct.a_set.iter();
+        let mut a_set_iter = state_reloaded.another_struct.a_set.iter().expect("Failed to get iterator");
```
Comment (abizjak) [inline examples/new-state/src/lib.rs:211]: In tests it is better to use `expect_report` since that will make `cargo concordium` report the error. This won't since panics are not reported, the program just terminates.

**Review by abizjak — CHANGES_REQUESTED** [review]

**Comment by Bargsteen** [comment]
Merged this into https://github.com/Concordium/concordium-rust-smart-contracts/tree/new-contract-state and resolved most of the comments.
I'll close this now, but keep the remaining comments as an additional todo list ;)
